File: android/app/src/main/java/com/example/cryptobenchmark/environment/discovery/AttributeKind.java
package com.example.cryptobenchmark.environment.discovery;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/** Single source of truth for attribute typing, shared by JCA parsing and JSON writing. */
public enum AttributeKind {
    LIST, INT, BOOL, STRING;

    private static final Map<String, AttributeKind> KINDS;

    static {
        Map<String, AttributeKind> kinds = new HashMap<String, AttributeKind>();
        kinds.put("SupportedModes", LIST);
        kinds.put("SupportedPaddings", LIST);
        kinds.put("SupportedKeyClasses", LIST);
        kinds.put("SupportedKeyFormats", LIST);
        kinds.put("SupportedCurves", LIST);
        kinds.put("KeySize", INT);
        kinds.put("ThreadSafe", BOOL);
        kinds.put("ImplementedIn", STRING);
        kinds.put("MechanismType", STRING);
        KINDS = Collections.unmodifiableMap(kinds);
    }

    public static AttributeKind of(String attribute) {
        AttributeKind kind = KINDS.get(attribute);
        return kind == null ? STRING : kind;
    }

    public static boolean isKnown(String attribute) {
        return KINDS.containsKey(attribute);
    }

    /** @throws NumberFormatException when an INT attribute does not hold a number. */
    public static Object parse(String attribute, String raw) {
        switch (of(attribute)) {
            case LIST:
                return splitPipe(raw);
            case INT:
                return Integer.valueOf(Integer.parseInt(raw.trim()));
            case BOOL:
                return Boolean.valueOf(Boolean.parseBoolean(raw.trim()));
            case STRING:
            default:
                return raw;
        }
    }

    public static List<String> splitPipe(String raw) {
        List<String> parts = new ArrayList<String>();
        if (raw == null) {
            return parts;
        }
        String[] pieces = raw.split("\\|");
        for (int i = 0; i < pieces.length; i++) {
            String trimmed = pieces[i].trim();
            if (!trimmed.isEmpty()) {
                parts.add(trimmed);
            }
        }
        return parts;
    }
}

File: environment/discovery/src/main/kotlin/adapter/AttributeKind.kt
package io.github.junekim0007.cryptobench.discovery.adapter

/** Single source of truth for attribute typing. */
internal enum class AttributeKind {
    LIST, INT, BOOL, STRING;

    companion object {

        private val KINDS = mapOf(
            "SupportedModes" to LIST,
            "SupportedPaddings" to LIST,
            "SupportedKeyClasses" to LIST,
            "SupportedKeyFormats" to LIST,
            "SupportedCurves" to LIST,
            "KeySize" to INT,
            "ThreadSafe" to BOOL,
            "ImplementedIn" to STRING,
            "MechanismType" to STRING,
        )

        fun of(attribute: String): AttributeKind = KINDS[attribute] ?: STRING

        /** @throws NumberFormatException when an INT attribute does not hold a number. */
        fun parse(attribute: String, raw: String): Any = when (of(attribute)) {
            LIST -> splitPipe(raw)
            INT -> raw.trim().toInt()
            BOOL -> raw.trim().toBoolean()
            STRING -> raw
        }

        fun splitPipe(raw: String): List<String> =
            raw.split('|').map { it.trim() }.filter { it.isNotEmpty() }
    }
}
